compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/CollectionStubMethodLowering.kt

```diff
@@ -0,0 +1,216 @@
+        var typeList = (targetClass.superTypes.filter { target ->
+            sourceClass.superTypes.any { it.classifierOrNull == target.classifierOrNull }
+        }.firstOrNull() as? IrSimpleType)?.arguments
+        if (typeList.isNullOrEmpty() || !typeList.all { it is IrType }) {
+            typeList = emptyList()
+        }
+        return sourceClass.typeParameters.map { it.symbol }.zip(typeList as List<IrType>).toMap()
+    }
+
+    // TODO: replace with new Type substitution utility
+    private fun IrType.substitute(substitutionMap: Map<IrTypeParameterSymbol, IrType>): IrType {
+        if (this !is IrSimpleType) return this
+
+        substitutionMap[classifier]?.let { return it }
+
+        val newArguments = arguments.map {
+            if (it is IrTypeProjection) {
+                makeTypeProjection(it.type.substitute(substitutionMap), it.variance)
+            } else {
+                it
+            }
+        }
+
+        val newAnnotations = annotations.map { it.deepCopyWithSymbols() }
+        return IrSimpleTypeImpl(
+            classifier,
+            hasQuestionMark,
+            newArguments,
+            newAnnotations
+        )
+    }
+
+    // Compute stubs that should be generated, compare based on signature
+    private fun computeStubsToGenerate(irClass: IrClass): HashSet<IrSimpleFunction> =
+        findRelevantSuperCollectionClasses(irClass).flatMap { (_, _, function) -> function.toHashSet() }.toHashSet()
+
+    private data class StubsForCollectionClass(
+        val readOnlyClass: IrClassSymbol,
+        val mutableClass: IrClassSymbol,
+        val mutableOnlyMethods: Collection<IrSimpleFunction>
+    )
+
+    private val preComputedStubs: Collection<StubsForCollectionClass> by lazy {
+        with(context.ir.symbols) {
+            listOf(
+                collection to mutableCollection,
+                set to mutableSet,
+                list to mutableList,
+                map to mutableMap,
+                mapEntry to mutableMapEntry,
+                iterable to mutableIterable,
+                iterator to mutableIterator,
+                listIterator to mutableListIterator
+            ).map { (readOnlyClass, mutableClass) ->
+                val mutableMethods = mutableListOf<IrSimpleFunction>()
+                val readOnlyMethodSignatures = readOnlyClass.functions.map { it.owner.toSignature() }.toHashSet()
+                for (mutableMethod in mutableClass.functions) {
+                    if (mutableMethod.owner.toSignature() !in readOnlyMethodSignatures) {
+                        mutableMethods.add(mutableMethod.owner)
+                    }
+                }
+                StubsForCollectionClass(readOnlyClass, mutableClass, mutableMethods)
+            }
+        }
+    }
+
+    private fun findRelevantSuperCollectionClasses(irClass: IrClass): Collection<StubsForCollectionClass> {
```

Comment:
In a hierarchy of Kotlin classes deriving from a collection, only the most general one requires stub methods. Thus, in
```
class K1: Map
class K2: K1
class K3: K2
```
only `K1` needs stubs.